backend/app/services/user_service.py

```
@@ -0,0 +1,32 @@
+class UserService:
+    def __init__(self, repo: UserRepo):
+        self.repo = repo
+
+    def signup(self, payload: UserCreateRequest) -> UserResponse:
+        # normalize inputs that require uniqueness
+        email = payload.email.strip().lower()
+        username = payload.username.strip().lower()
+
+        # pre-check for DB uniqueness, still enforced in repo
+        if self.repo.get_by_email(email):
+            raise EmailInUseError("email already in use")
```

Ruff (0.13.2)

Avoid specifying long messages outside the exception class

(TRY003)